sync2/atxs.go

```diff
@@ -0,0 +1,326 @@
+	"github.com/spacemeshos/go-spacemesh/sync2/multipeer"
+	"github.com/spacemeshos/go-spacemesh/sync2/rangesync"
+	"github.com/spacemeshos/go-spacemesh/sync2/sqlstore"
+	"github.com/spacemeshos/go-spacemesh/system"
+)
+
+const (
+	proto = "sync/2"
+)
+
+type ATXHandler struct {
+	logger           *zap.Logger
+	f                Fetcher
+	clock            clockwork.Clock
+	batchSize        int
+	maxAttempts      int
+	maxBatchRetries  int
+	failedBatchDelay time.Duration
+}
+
+var _ multipeer.SyncKeyHandler = &ATXHandler{}
+
+func NewATXHandler(
+	logger *zap.Logger,
+	f Fetcher,
+	batchSize, maxAttempts, maxBatchRetries int,
+	failedBatchDelay time.Duration,
+	clock clockwork.Clock,
+) *ATXHandler {
+	if clock == nil {
+		clock = clockwork.NewRealClock()
+	}
+	return &ATXHandler{
+		f:                f,
+		logger:           logger,
+		clock:            clock,
+		batchSize:        batchSize,
+		maxAttempts:      maxAttempts,
+		maxBatchRetries:  maxBatchRetries,
+		failedBatchDelay: failedBatchDelay,
+	}
+}
+
+func (h *ATXHandler) Commit(
+	ctx context.Context,
+	peer p2p.Peer,
+	base rangesync.OrderedSet,
+	received rangesync.SeqResult,
+) error {
+	h.logger.Debug("begin atx commit")
+	defer h.logger.Debug("end atx commit")
+	var firstK rangesync.KeyBytes
+	numDownloaded := 0
+	state := make(map[types.ATXID]int)
+	for k := range received.Seq {
+		if firstK == nil {
+			firstK = k
+		} else if firstK.Compare(k) == 0 {
+			break
+		}
+		found, err := base.Has(k)
+		if err != nil {
+			return fmt.Errorf("check if ATX exists: %w", err)
+		}
+		if found {
+			continue
+		}
+		id := types.BytesToATXID(k)
+		h.f.RegisterPeerHash(peer, id.Hash32())
+		state[id] = 0
+	}
+	if err := received.Error(); err != nil {
+		return fmt.Errorf("get item: %w", err)
+	}
+	total := len(state)
+	items := make([]types.ATXID, 0, h.batchSize)
+	startTime := h.clock.Now()
+	batchAttemptsRemaining := h.maxBatchRetries
+	for len(state) > 0 {
+		items = items[:0]
+		for id, n := range state {
+			if n >= h.maxAttempts {
+				h.logger.Debug("failed to download ATX: max attempts reached",
+					zap.String("atx", id.ShortString()))
+				delete(state, id)
+				continue
+			}
+			items = append(items, id)
+			if len(items) == h.batchSize {
+				break
+			}
+		}
+		if len(items) == 0 {
+			break
+		}
+
+		someSucceeded := false
+		var mtx sync.Mutex
+		err := h.f.GetAtxs(ctx, items, system.WithATXCallback(func(id types.ATXID, err error) {
+			mtx.Lock()
+			defer mtx.Unlock()
+			switch {
+			case err == nil:
+				numDownloaded++
+				someSucceeded = true
+				delete(state, id)
+			case errors.Is(err, pubsub.ErrValidationReject):
+				// if the atx invalid there's no point downloading it again
+				state[id] = h.maxAttempts
+			default:
+				state[id]++
+			}
+		}))
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				return err
+			}
+			if !errors.Is(err, &fetch.BatchError{}) {
+				h.logger.Debug("failed to download ATXs", zap.Error(err))
+			}
+		}
```

Comment:
NIT:

```suggestion
		switch {
		case err == nil:
		case errors.Is(err, context.Canceled):
			return err
		case !errors.Is(err, &fetch.BatchError{}):
			h.logger.Debug("failed to download ATXs", zap.Error(err))
		}
```

also what if `err` is a `&fetch.BatchError{}` do we then just ignore it?

Reply:
We ignore `fetch.BatchError` b/c in this case the errors were already handled in the callbacks

Reply:
Applied the patch